### PR TITLE
Music room Streaming and Listening

### PR DESCRIPTION
**Pull request by uuilmer**
Streamers:
1. Allow people to Stream their Spotify when clicked a button in their profile.
1. Can stop streaming by clicking button in their own profile.

Listeners:
1. Allow people to listen to a friend's stream when clicked a button in the friend's profile.
1. Can stop listening without having to look for what friend their are listening to: click button in your own profile.

Demonstration:

ezgif-6-f706ef77e6fc

### CODE REVIEW

File: app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java [hunk 1]
```diff
@@ -101,6 +102,12 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
             Toast.makeText(getContext(), "Error retrieving more chats", Toast.LENGTH_SHORT).show();
         }
 
+        ivProfile.setOnClickListener(view12 -> getParentFragmentManager().beginTransaction()
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java:106]: Why view12 though? Why not view11 or view13?

File: app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java [hunk 2]
```diff
@@ -101,6 +102,12 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
+                .setCustomAnimations(R.anim.slide_in_up, R.anim.slide_out_up)
+                .addToBackStack("stack")
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java:108]: What is "stack"?

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    public static Timer listenerTimer;
+    private ImageView ivPlay;
+    // We don't need to worry about making the remote static and limiting that only one playerApi can be subscribed to,
+    // because the only person that can subscribe to a playerapi is the owner, who can only do it in their
+    // profile screen.
+    private SpotifyAppRemote remote;
+    private ParseQuery<MusicRoom> q;
+    public static boolean isOwner;
+    private ImageView playingGif;
+    private final MusicRoom[] musicRoom = new MusicRoom[1];
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: Why the array?

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        // Try to get the current Song
+        nowPlayingInParse[0] = musicRoom[0].getCurrentSong();
+
+        // Case where the current user needs to create a Song for this MusicRoom(None playing)
+        if (nowPlayingInParse[0] == null) {
+            seekSong();
+        }
+        // Case where there is already a Song playing (This is mainly for listeners)
+        else
+            setup(musicRoom);
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: else {...}

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+                    } catch (ParseException e) {
+                        e.printStackTrace();
+                        Toast.makeText(getContext(), "Error looking for new Song!", Toast.LENGTH_SHORT).show();
+                        this.cancel();
+                    }
+                }
+            }, 0, 50);
+    }
+
+    final boolean[] isPlayingLocally = new boolean[1];
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: Move this to the top please.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                ivPlay.setImageResource(R.drawable.musicroom_play);
+
+            ivPlay.setOnClickListener(view1 -> startSync(musicRoom));
+        } catch (ParseException e) {
+            e.printStackTrace();
+            Toast.makeText(getContext(), "Cannot set up Stream!", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    final Song[] nowPlayingInParse = new Song[1];
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: Same.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java (hunk 5 above)
Comment (xuyueee) [reply app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: Also why array?

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+                musicRoom[0].saveInBackground(e1 -> {
+                    if (e1 != null) {
+                        Toast.makeText(getContext(), "Error updating MusicRoom!", Toast.LENGTH_SHORT).show();
+                        return;
+                    }
+                    setup(musicRoom);
+                });
+            });
+            // If the user is a listener and there is no Song, they must wait until the MusicRoom has a Song
+        } else
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java:134]: curly braces please.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+            nowPlayingInParse[0] = new Song();
+            nowPlayingInParse[0].setPlaying(false);
+            nowPlayingInParse[0].saveInBackground(e -> {
+                if (e != null) {
+                    Toast.makeText(getContext(), "Error starting song!", Toast.LENGTH_SHORT).show();
+                    return;
+                }
+                // The MusicRoom needs to point at this new Song
+                musicRoom[0].setCurrentSong(nowPlayingInParse[0]);
+                musicRoom[0].saveInBackground(e1 -> {
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: Tip: `e1` can still be `e`, since they are not in the same scope.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+        if (!isPlayingLocally[0]) {
+            ivPlay.setImageResource(R.drawable.musicroom_pause);
+            playingGif.setVisibility(View.VISIBLE);
+            try {
+                musicRoom[0] = q.getFirst();
+                // Fetch the song that's playing at the moment
+                nowPlayingInParse[0] = musicRoom[0].getCurrentSong().fetch();
+
+                // Case where the User is the owner
+                final boolean[] isPlayingInParse = {nowPlayingInParse[0].isPlaying()};
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: Can you make this a member variable?

**Comment by xuyueee** [comment]
Could you attach a screenshot or video for this feature? It's hard to tell how it works just from reading the code.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+        // Case where there is already a Song playing (This is mainly for listeners)
+        else
+            setup(musicRoom);
+    }
+
+    // Method where this MusicRoom has no Song, and we therefore need to create one
+    private void seekSong() {
+        // If this User is the owner of the room, he can create on himself and move on
+        if (isOwner) {
+            nowPlayingInParse[0] = new Song();
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: How did this achieve finding a song? Do we know what song it is at this point?

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+            ivPlay.setOnClickListener(view1 -> startSync(musicRoom));
+        } catch (ParseException e) {
+            e.printStackTrace();
+            Toast.makeText(getContext(), "Cannot set up Stream!", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    final Song[] nowPlayingInParse = new Song[1];
+
+    private void startSync(final MusicRoom[] musicRoom) {
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: I think this method is too long and hard to follow with all the if else statements. You can either split it up by owner or listener, or by playing or paused.

File: app/src/main/java/com/example/pix/home/fragments/ProfileFragment.java [hunk 12]
```diff
@@ -34,8 +36,11 @@
 
 public class ProfileFragment extends Fragment {
 
-    public ProfileFragment() {
-        // Required empty public constructor
+    ParseUser user;
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/home/fragments/ProfileFragment.java]: private

File: app/src/main/java/com/example/pix/home/fragments/ProfileFragment.java [hunk 13]
```diff
@@ -46,74 +51,94 @@ public void onCreate(Bundle savedInstanceState) {
...
+            });
+        } else {
+            // The friend's name is a TextView, so that the user cannot edit it
+            TextView name = view.findViewById(R.id.profile_name);
+            name.setText("" + user.getUsername());// If we press "Enter" in the username EditText, update the username
+        }
         // Set User's number of Pix
         TextView pix = view.findViewById(R.id.profile_pix);
-        pix.setText("" + ParseUser.getCurrentUser().getInt("pix"));
+        pix.setText("" + user.getInt("pix"));
```
Comment (xuyueee) [inline app/src/main/java/com/example/pix/home/fragments/ProfileFragment.java]: Constant please.

**Review by xuyueee — COMMENTED** [review]
Glad to see your first draft of the implementation of a fairly complex feature! The comments about coding style from my review of the last pull request still apply to this one. I've pointed out some places here, but please apply the suggestions to the whole pull request.

File: app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java (hunk 1 above)
Comment (uuilmer) [reply app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java:106]: It seems like this is the same typo as the previous pull request, and was corrected by renaming view12(Which was a typo) into unusedView. Do I just fix the issues again?

File: app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java (hunk 2 above)
Comment (uuilmer) [reply app/src/main/java/com/example/pix/chat/fragments/ChatFragment.java:108]: I actually removed this line in a future pull request. I attempted to add the current fragment (Chatting with a friend) to a back stack I just named "stack", then going to the profilefragment of this friend, and when Im done retrieving the chat fragment in my stack. I decided to change it but now that I think about it, I will try to do this again, since the way I have it now retrieves the previous fragment by making it a public static variable.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java (hunk 3 above)
Comment (uuilmer) [reply app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: Same error as before, where I didnt know that I can just make them member variables to avoid making them a final single member array.

File: app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java (hunk 10 above)
Comment (uuilmer) [reply app/src/main/java/com/example/pix/chat/fragments/MusicRoomFragment.java]: This method is called from the onCreateView method whenever we find that the Song we found from Parse is null. If this USer is the owner of the MusicRoom, he can create a new Song and save it to Parse. The else is for people who join a MusicRoom with no song. I handled them by creating a Timer that keeps checking until a Song is created in Parse by the owner for this MusicRoom.
